src/cc/table_desc.h

```diff
@@ -16,18 +16,101 @@
 
 #pragma once
 
+#include <unistd.h>
 #include <cstdint>
+#include <memory>
 #include <string>
 
 namespace llvm {
 class Function;
 }
 
+namespace clang {
+class ASTContext;
+class QualType;
+}
+
 namespace ebpf {
 
-struct TableDesc {
-  std::string name;
+class TableDesc;
+
+/// FileDesc is a helper class for managing open file descriptors. Copy is
+/// disallowed (call dup instead), and cleanup happens automatically.
+class FileDesc {
+  friend TableDesc;
+
+ private:
+  FileDesc &operator=(const FileDesc &that) {
+    fd = ::dup(that.fd);
+    return *this;
+  }
+  FileDesc(const FileDesc &that) { *this = that; }
+
+ public:
+  FileDesc(int fd = -1) : fd(fd) {}
+  FileDesc &operator=(FileDesc &&that) {
+    fd = that.fd;
+    that.fd = -1;
+    return *this;
+  }
+  FileDesc(FileDesc &&that) { *this = std::move(that); }
+  ~FileDesc() {
+    if (fd >= 0)
+      ::close(fd);
+  }
+  FileDesc dup() const { return FileDesc(*this); }
+
+  operator int() { return fd; }
+  operator int() const { return fd; }
+
+ private:
   int fd;
+};
+
+/// TableDesc uniquely stores all of the runtime state for an active bpf table.
+/// The copy constructor/assign operator are disabled since the file handles
+/// owned by this table are not implicitly copyable. One should call the dup()
+/// method if an explicit new handle is required. We define the move operators
+/// so that objects of this class can reside in stl containers.
+class TableDesc {
+ private:
+  TableDesc(const TableDesc &) = default;
+  TableDesc &operator=(const TableDesc &) = default;
+
+ public:
+  TableDesc()
```

Comment:
`fd` is not initialized for the default empty constructor. What's the behavior if it's actually being used? Since we have fall-back like `BPFHashTable<>({})` which uses this......